README
NAME
    Test::CompanionClasses - Basic invocation of the test companion class
    engine

VERSION
    version 1.100870

SYNOPSIS
        # Define a test file, for example C<t/01_companion_classes.t>:

        use Test::CompanionClasses;
        run_tests;

        # Then you can do:
        # perl t/01_companion_classes.t --exact Foo::Bar Baz

DESCRIPTION
    This is a very basic frontend for Test::CompanionClasses::Engine which
    you can use for your distribution test files (in "t/").

    The intention is that you use it as shown in the "SYNOPSIS".

METHODS
  run_tests
    Parses the command-line options, then calls the "run_tests()" method of
    Test::CompanionClasses::Engine.

COMMAND-LINE USAGE
    The following command-line arguments are supported:

    --exact
        Specifies that the package filter is to be used exactly, i.e.,
        substring matching is not enough. See Test::CompanionClasses::Engine
        for details.

    The rest of the command line is interpreted as a list of package
    filters. Again, see Test::CompanionClasses::Engine for details.

    The "inherited" mechanism is not supported (yet).

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Test-CompanionClasses>
    .

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Test-CompanionClasses/>.

    The development version lives at
    <http://github.com/hanekomu/Test-CompanionClasses/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Test-CompanionClasses

1.100870  2010-03-28 12:14:51 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.06  Mon 2009.11.02 19:38:28 CET (Marcel Gruenauer <marcel@cpan.org>)
    - subtest() requires Test::More 0.94 (thanks ANDK)

0.05  Mon 2009.11.02 19:38:28 CET (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.
    - Use Test::More's subtest feature - this greatly simplified the engine.

0.04  Thu, 29 May 2008 12:08:27 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - set the version to 0.04
    - moved generated documentation of interited methods to after documentation
      of this class' methods so the manpage is more readable
    - .shipit: fixed svk.tagpattern

0.03  Wed, 21 Nov 2007 18:23:41 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - Converted Changes file to YAML style
    - .shipit: added Twitter step
    - Makefile.PL: added auto_install() and process_templates()
    - lib/*: converted to template
    - updated MANIFEST
    - Test::CompanionClasses::Base: make_real_object() now passes args to the
      real object's constructor
    - updated t/perlcriticrc
    - tags: NEWFEATURE

0.02  Thu, 25 Oct 2007 22:23:31 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - Makefile.PL now requires Data::Inherited

0.01  Thu, 25 Oct 2007 15:34:05 +0200 (Marcel Gruenauer <marcel@cpan.org>)
    - original version


lib/Test/CompanionClasses.pm
package Test::CompanionClasses;
# ABSTRACT: Basic invocation of the test companion class engine
use 5.006;
use warnings;
use strict;
use Test::CompanionClasses::Engine;
use Getopt::Long;
use Exporter qw(import);
our $VERSION = '0.06';
our @EXPORT  = ('run_tests');

sub run_tests {
    my $exact;
    GetOptions(exact => \$exact)
      or die "usage: $0 [ --exact ] filter...\n";
    Test::CompanionClasses::Engine->new->run_tests(
        exact  => $exact,
        filter => [@main::ARGV],

        # inherited => [ $inherited_spec ],
    );
}
1;


__END__
=pod

=head1 NAME

Test::CompanionClasses - Basic invocation of the test companion class engine

=head1 VERSION

version 1.100870

=head1 SYNOPSIS

    # Define a test file, for example C<t/01_companion_classes.t>:

    use Test::CompanionClasses;
    run_tests;

    # Then you can do:
    # perl t/01_companion_classes.t --exact Foo::Bar Baz

=head1 DESCRIPTION

This is a very basic frontend for L<Test::CompanionClasses::Engine> which you
can use for your distribution test files (in C<t/>).

The intention is that you use it as shown in the L</SYNOPSIS>.

=head1 METHODS

=head2 run_tests

Parses the command-line options, then calls the C<run_tests()> method of
L<Test::CompanionClasses::Engine>.

=head1 COMMAND-LINE USAGE

The following command-line arguments are supported:

=over 4

=item --exact

Specifies that the package filter is to be used exactly, i.e., substring
matching is not enough. See L<Test::CompanionClasses::Engine> for details.

=back

The rest of the command line is interpreted as a list of package filters.
Again, see L<Test::CompanionClasses::Engine> for details.

The C<inherited> mechanism is not supported (yet).

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Test-CompanionClasses>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Test-CompanionClasses/>.

The development version lives at
L<http://github.com/hanekomu/Test-CompanionClasses/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Test/CompanionClasses/Base.pm
use 5.008;
use strict;
use warnings;

package Test::CompanionClasses::Base;
our $VERSION = '1.100870';
# ABSTRACT: Base class for test companion classes
use Test::More;
use UNIVERSAL::require;
use parent qw(
  Class::Accessor::Complex
  Data::Inherited
);
__PACKAGE__
    ->mk_new
    ->mk_scalar_accessors(qw(package));
use constant PLAN => 0;    # default

sub make_real_object {
    my ($self, @args) = @_;
    $self->package->require;
    die $@ if $@;
    $self->package->new(@args);
}
sub run { }

# this can be called as a class method as well
sub planned_test_count {
    my $self = shift;
    my $plan = 0;
    $plan += $_ for $self->every_list('PLAN');
    $plan;
}
1;


__END__
=pod

=for test_synopsis my ($foo, $bar);

=head1 NAME

Test::CompanionClasses::Base - Base class for test companion classes

=head1 VERSION

version 1.100870

=head1 SYNOPSIS

    package My::Foo_TEST;

    use warnings;
    use strict;

    use base 'Test::CompanionClasses::Base';

    use constant PLAN => 5;

    sub run {
        my $self = shift;
        $self->SUPER::run(@_);
        is_deeply($foo, $bar, 'some test');
        # ...
    }

=head1 DESCRIPTION

Base class for test companion classes. Each test companion class should
inherit from this class.

The package() property is automatically set; it holds the package name of the
class you are testing. If your test companion class is called C<My::Foo_TEST>,
then C<package()> will return C<My::Foo>.

=head1 METHODS

=head2 PLAN

A constant that says how many tests this particular class defines. Real test
companion classes (i.e., subclasses of this class) will want to redefine it
like this:

    use constant PLAN => 5;

Note that you should only specify how many tests the current class runs; test
counts of superclasses are automatically taken care of.

=head2 planned_test_count

Uses C<PLAN()>, calculated over the test companion class' whole class
hierarchy, to determine how many tests will be run in total.

=head2 make_real_object

Loads the actual class being tested (see C<package()>) and returns an object
of this class (constructed by calling C<new()> on it).

In your test companion class you will want to test certain assumptions about
your real class, so this method will be useful.

=head2 run

Test companion classes should override this method and run their tests. Be
sure to call C<SUPER::run(@_)> so that all tests over the class hierarchy are
run.

The C<run()> method in this base class just prints a line informing the test
user that tests for this particular companion class have begun. If you have
several companion classes - and you probably will or you won't have been using
C<Test::CompanionClasses> - this serves as a visual distinction of where on
companion class' tests end the next ones' begin.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Test-CompanionClasses>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Test-CompanionClasses/>.

The development version lives at
L<http://github.com/hanekomu/Test-CompanionClasses/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Test/CompanionClasses/Engine.pm
use 5.008;
use strict;
use warnings;

package Test::CompanionClasses::Engine;
our $VERSION = '1.100870';
# ABSTRACT: Run tests defined in companion classes
use FindBin '$Bin';
use File::Find;
use Test::More;
use UNIVERSAL::require;
use parent 'Class::Accessor::Complex';
__PACKAGE__->mk_new;

sub run_tests {
    my ($self, %args) = @_;
    our $package_filter = join '|' => map { "\Q$_\E" } @{ $args{filter} || [] };
    $package_filter = "^($package_filter)\$" if $args{exact};

    # default uses lib/ one level up, as the program using this class usually
    # lives in t/
    my $lib = $args{lib} || "$Bin/../lib";
    my @test_packages =
      map { { real_package => $_, test_package => $_ . '_TEST', } }
      grep { $package_filter ? m/$package_filter/ : 1 }
      @{ $args{inherited} || [] };
    find(
        sub {
            return unless -f && /_TEST\.pm$/;
            my $test_package;
            if ($File::Find::name =~ m!\Q$lib\E/(.*_TEST)\.pm$!) {
                ($test_package = $1) =~ s!/!::!g;
            } else {
                die "can't determine package from filename [$File::Find::name]";
            }
            (my $real_package = $test_package) =~ s/_TEST$//;
            return if $package_filter && ($real_package !~ $package_filter);
            push @test_packages => {
                real_package => $real_package,
                test_package => $test_package,
            };
        },
        $lib
    );
    plan tests => scalar @test_packages;
    for (@test_packages) {
        subtest $_->{test_package}, sub {
            $_->{test_package}->require;
            die $@ if $@;
            plan tests => $_->{test_package}->planned_test_count;
            $_->{test_package}->new('package' => $_->{real_package})->run;
        };
    }
}
1;


__END__
=pod

=head1 NAME

Test::CompanionClasses::Engine - Run tests defined in companion classes

=head1 VERSION

version 1.100870

=head1 SYNOPSIS

    use Test::CompanionClasses;
    Test::CompanionClasses->new->run_tests('...');

=head1 DESCRIPTION

This is the core of C<Test::CompanionClasses>.

=head1 METHODS

=head2 run_tests

Actually runs the companion class tests.

Takes named arguments (as a hash). Recognized keys are:

=over 4

=item C<filter>

A reference to a list of strings that are interpreted as package filters. A
companion test class is only run if the corresponding real class' package name
matches this filter list.

=item C<exact>

Works with C<filter>. If this boolean flag is set, the real class name must
match exactly, otherwise a substring match is sufficient.

Examples:

    Test::CompanionClasses->mk_new->run_tests(
        filter => [ qw/Foo::Bar Baz/ ]
    );

will run the companion tests of C<Foo::Bar>, C<Baz> but also
C<Foo::Bar::Flurble>, C<Bazzzz> etc.

    Test::CompanionClasses->mk_new->run_tests(
        filter => [ qw/Foo::Bar Baz/ ]
        exact  => 1,
    );

will only run the companion tests of C<Foo::Bar> and C<Baz>.

=item C<lib>

Sets the directory in and under which C<run_tests()> is looking for test
companion classes. Defaults to C<$Bin/../lib>, where C<$Bin> is the location
of the program as determined by L<FindBin>. This default is used because
normally companion class tests will be run from a perl distribution's C<t/>
directory.

=item C<inherited>

You can also specify that other classes not found in C<lib> should be tested.
Use a reference to an array of class names as the value for C<inherited> and
those classes' companion tests will be run as well. The class names still have
to match the C<filter>, if one was specified.

This is useful if your distribution depends on another one which also has
defined test companion classes. If your distribution changes the way these
other test companion classes are working, you can inherit those tests to see
whether they still work with your distribution.

=back

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Test-CompanionClasses>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Test-CompanionClasses/>.

The development version lives at
L<http://github.com/hanekomu/Test-CompanionClasses/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Test::CompanionClasses',
                       'DISTNAME' => 'Test-CompanionClasses',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Basic invocation of the test companion class engine',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100870',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Getopt::Long' => '0',
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'parent' => '0',
                                        'FindBin' => '0',
                                        'constant' => '0',
                                        'Exporter' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'UNIVERSAL::require' => '0',
                                        'Data::Inherited' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



